postgres_exporter/common/pg16/setup.sql

```diff
@@ -0,0 +1,463 @@
+
+
+DROP FUNCTION IF EXISTS monitor.pg_settings_checksum(text);
+CREATE FUNCTION monitor.pg_settings_checksum(p_known_settings_hash text DEFAULT NULL) 
+    RETURNS smallint
+    LANGUAGE plpgsql SECURITY DEFINER 
+    SET search_path TO pg_catalog, pg_temp
+AS $function$
+DECLARE
+
+v_is_in_recovery        boolean;
+v_settings_hash         text;
+v_settings_hash_old     text;
+v_settings_match        smallint := 0;
+v_settings_string       text;
+v_settings_string_old   text;
+v_valid                 smallint;
+
+BEGIN
+
+SELECT pg_is_in_recovery() INTO v_is_in_recovery;
+
+SELECT md5_hash
+    , settings_string
+INTO v_settings_hash
+    , v_settings_string
+FROM monitor.pg_settings_hash;
+
+SELECT settings_hash_generated, valid
+INTO v_settings_hash_old, v_valid
+FROM monitor.pg_settings_checksum
+ORDER BY created_at DESC LIMIT 1;
+
+IF p_known_settings_hash IS NOT NULL THEN
+    v_settings_hash_old := p_known_settings_hash;
+    -- Do not base validity on the stored value if manual hash is given. 
+    v_valid := 0;
+END IF;
+
+IF (v_settings_hash_old IS NOT NULL) THEN
+
+    IF (v_settings_hash != v_settings_hash_old) THEN
+
+        v_valid := 1;
+
+        IF v_is_in_recovery = false THEN 
+            INSERT INTO monitor.pg_settings_checksum (
+                    settings_hash_generated
+                    , settings_hash_known_provided
+                    , settings_string
+                    , valid)
+            VALUES (
+                    v_settings_hash
+                    , p_known_settings_hash
+                    , v_settings_string
+                    , v_valid);
+        END IF;
+    END IF;
+
+ELSE
+
+    v_valid := 0;
+    IF v_is_in_recovery = false THEN
+        INSERT INTO monitor.pg_settings_checksum (
+                settings_hash_generated
+                , settings_hash_known_provided
+                , settings_string
+                , valid)
+        VALUES (v_settings_hash
+                , p_known_settings_hash
+                , v_settings_string
+                , v_valid);
+    END IF;
+
+END IF; 
+
+RETURN v_valid;
+
+END
+$function$;
+
+
+DROP FUNCTION IF EXISTS monitor.pg_hba_checksum(text);
+CREATE FUNCTION monitor.pg_hba_checksum(p_known_hba_hash text DEFAULT NULL) 
+    RETURNS smallint
+    LANGUAGE plpgsql SECURITY DEFINER 
+    SET search_path TO pg_catalog, pg_temp
+AS $function$
+DECLARE
+
+v_hba_hash              text;
+v_hba_hash_old          text;
+v_hba_match             smallint := 0;
+v_hba_string            text;
+v_hba_string_old        text;
+v_is_in_recovery        boolean;
+v_valid                 smallint;
+
+BEGIN
+
+SELECT pg_is_in_recovery() INTO v_is_in_recovery;
+
+IF current_setting('server_version_num')::int >= 100000 THEN
+
+    SELECT md5_hash
+        , hba_string
+    INTO v_hba_hash
+        , v_hba_string
+    FROM monitor.pg_hba_hash;
+
+ELSE
+    RAISE EXCEPTION 'pg_hba change monitoring unsupported in versions older than PostgreSQL 10';
+END IF;
```

Comment:
can' this be dropped? PG12+ is all the community support now, right?

Reply:
11 still has support for one more patch release, so mid-November.

Reply:
That said, this specific thing may be fine, just speaking to the "supported versions"—as well, no need to artificially remove support if things could still work, IMHO.